src/NuGet.Core/NuGet.Packaging/Signing/Verification/AllowListVerificationProvider.cs

```diff
@@ -30,32 +20,88 @@ public Task<PackageVerificationResult> GetTrustResultAsync(ISignedPackageReader
             }
 
             return new SignedPackageVerificationResult(status, signature, issues);
         }
 
-        private bool IsSignatureAllowed(PrimarySignature signature)
+        private SignatureVerificationStatus VerifyAllowList(
+            Signature signature,
+            List<SignatureLog> issues,
+            IReadOnlyList<VerificationAllowListEntry> allowList,
+            bool requireAllowList,
+            bool issuesAreFatal,
+            string noListErrorMessage,
+            string noMatchErrorMessage)
         {
-            // Get information needed for allow list verification
-            var primarySignatureCertificateFingerprint = CertificateUtility.GetHash(signature.SignerInfo.Certificate, _fingerprintAlgorithm);
-            var primarySignatureCertificateFingerprintString = BitConverter.ToString(primarySignatureCertificateFingerprint).Replace("-", "");
+            var status = SignatureVerificationStatus.Valid;
 
-            foreach (var allowedEntry in _allowList)
+            if (allowList == null || allowList.Count == 0)
+            {
+                if (requireAllowList)
+                {
+                    status = SignatureVerificationStatus.Untrusted;
+                    issues.Add(SignatureLog.Issue(fatal: issuesAreFatal, code: NuGetLogCode.NU3034, message: noListErrorMessage));
+                }
+            }
+            else if (!IsSignatureAllowed(signature, allowList))
+            {
+                status = SignatureVerificationStatus.Untrusted;
+                issues.Add(SignatureLog.Issue(fatal: issuesAreFatal, code: NuGetLogCode.NU3034, message: noMatchErrorMessage));
+            }
+
+            return status;
+        }
+
+        private bool IsSignatureAllowed(
+            Signature signature,
+            IReadOnlyList<VerificationAllowListEntry> allowList)
+        {
+            var target = VerificationTarget.Primary;
+
+            if (signature.Type == SignatureType.Repository)
+            {
+                target = VerificationTarget.Repository;
+            }
+
+            foreach (var allowedEntry in allowList)
             {
                 // Verify the certificate hash allow list objects
                 var certificateHashEntry = allowedEntry as CertificateHashAllowListEntry;
                 if (certificateHashEntry != null)
                 {
-                    if (certificateHashEntry.VerificationTarget.HasFlag(VerificationTarget.Primary) &&
-                        StringComparer.OrdinalIgnoreCase.Equals(certificateHashEntry.CertificateFingerprint, primarySignatureCertificateFingerprintString))
+                    // Get information needed for allow list verification
+                    var primarySignatureCertificateFingerprint = CertificateUtility.GetHash(signature.SignerInfo.Certificate, certificateHashEntry.FingerprintAlgorithm);
```

Comment:
can you memoize them at least? We shouldn't be calculating the same hash more than once. If the entries have different hash algorithms we should only calculate the hash once per hash algorithm. You can create a HashMap where the key is the HashAlgorithmName and the value is the fingerprint and use that to avoid calculating it multiple times